### PR TITLE
Parameters are now associated to section in the parameter parser

### LINKED ISSUE #256: In generaldata.ini, section name is ignored
### Description of the issue
The two following **generaldata.ini** files are interpreted in the same way, hence they're equivalent :

> [general]
> mode = Economy

> [output]
> mode = Economy

This could lead to conflicts if two sections contained the same property : `section1/property` and `section2/property` would not be distinguished.

## Possible fix
In order to fix this, the section name should to be taken into account in source file **parameters.cpp** by adding this parameter to the relevant functions.

### PR DESCRIPTION
**Pull request by julza03**
The SGDIntLoadFamily_* functions, where * was a letter, were replaced… by functions where * correspond to a section name.
Thus, parameters are associated to their corresponding section.
Parameters kept for compatibility reasons are grouped in a single function

### CODE REVIEW

File: src/libs/antares/study/parameters.cpp [hunk 1]
```diff
@@ -691,327 +791,169 @@ static bool SGDIntLoadFamily_P(Parameters& d, const String& key, const String& v
...
-    }
-    else
-    {
-        int i = c - 'a';
-        if (jumper[i])
-            return (*(jumper[i]))(params, key, value, version);
-    }
-    return false;
+
+
```
Comment (flomnes) [inline src/libs/antares/study/parameters.cpp]: Rename this function to `checkKeySection(const String& s)` to check that the 1st character is in range a-z.

Call it whenever a section/property name is read.

File: src/libs/antares/study/parameters.cpp [hunk 2]
```diff
@@ -1020,10 +962,33 @@ bool Parameters::loadFromINI(const IniFile& ini, uint version, const StudyLoadOp
...
+    static std::map<String,Callback> sectionAssociatedToKeysProcess = {
+            {"general",&SGDIntLoadFamily_General},
+            {"input",&SGDIntLoadFamily_Input},
+            {"output",&SGDIntLoadFamily_Output},
+            {"optimization",&SGDIntLoadFamily_Optimization},
+            {"other preferences",&SGDIntLoadFamily_OtherPreferences},
+            {"advanced parameters",&SGDIntLoadFamily_AdvancedParameters},
+            {"playlist",&SGDIntLoadFamily_Playlist},
+            {"variables selection",&SGDIntLoadFamily_VariablesSelection},
+            {"seeds",&SGDIntLoadFamily_Seeds}};
```
Comment (flomnes) [inline src/libs/antares/study/parameters.cpp]: The complete name is **seeds - mersenne twister**

File: src/libs/antares/study/parameters.cpp [hunk 3]
```diff
@@ -1020,10 +962,33 @@ bool Parameters::loadFromINI(const IniFile& ini, uint version, const StudyLoadOp
...
+
+    Callback* handleAllKeysInSection;
```
Comment (flomnes) [inline src/libs/antares/study/parameters.cpp]: `Callback` is already a function pointer. This is equivalent to 
```cpp
function_t** Callback;
```
I suggest you remove one indirection level.

File: src/libs/antares/study/parameters.cpp [hunk 4]
```diff
@@ -1020,10 +962,33 @@ bool Parameters::loadFromINI(const IniFile& ini, uint version, const StudyLoadOp
...
     // Foreach section on the ini file...
     for (auto* section = ini.firstSection; section; section = section->next)
     {
+        sectionName = section->name;
+        sectionName.toLower();
+        handleAllKeysInSection = &sectionAssociatedToKeysProcess[section->name]; // @TODO Must check that the function exists
```
Comment (flomnes) [inline src/libs/antares/study/parameters.cpp]: 1. Use `std::map<T>::count` to determine the existence of the key within the map.
2. Use `sectionName` for the key.
3. Use `std::map<T>::at`, it won't try to add elements to your map unlike `operator[]`. More details on [at](https://www.cplusplus.com/reference/map/map/at/) and [operator\[\]](https://www.cplusplus.com/reference/map/map/operator[]/).

File: src/libs/antares/study/parameters.cpp [hunk 5]
```diff
@@ -1783,7 +1756,7 @@ void Parameters::saveToINI(IniFile& ini) const
 
     // Seeds
     {
-        auto* section = ini.addSection("seeds - Mersenne Twister");
+        auto* section = ini.addSection("seeds - Mersenne Twister"); //@TODO : Why Capital letters are used here ?
```
Comment (flomnes) [inline src/libs/antares/study/parameters.cpp]: I suggest you keep this as is for compatibility reasons.

Possible explanation [here](https://fr.wikipedia.org/wiki/Mersenne_Twister).

File: src/libs/antares/study/parameters.cpp [hunk 6]
```diff
@@ -1020,10 +962,33 @@ bool Parameters::loadFromINI(const IniFile& ini, uint version, const StudyLoadOp
     reset();
     // A temporary buffer, used for the values in lowercase
     String value;
-
+    String sectionName;
+    typedef bool (*Callback)(Parameters&, const String&, const String&, const String&, const String&, uint);
+
+    static std::map<String,Callback> sectionAssociatedToKeysProcess = {
```
Comment (flomnes) [inline src/libs/antares/study/parameters.cpp]: `static const`

**Review by flomnes — CHANGES_REQUESTED** [review]
Thank you for your contribution !

File: src/libs/antares/study/parameters.cpp (hunk 6 above)
Comment (julza03) [reply src/libs/antares/study/parameters.cpp]: OK. Commit 55c18c4

File: src/libs/antares/study/parameters.cpp (hunk 5 above)
Comment (julza03) [reply src/libs/antares/study/parameters.cpp]: OK. See [commit ](https://github.com/AntaresSimulatorTeam/Antares_Simulator/pull/342/commits/a5a76f731abb174f417a1e70ec7f1fd996da2981)

File: src/libs/antares/study/parameters.cpp (hunk 4 above)
Comment (julza03) [reply src/libs/antares/study/parameters.cpp]: Ok. Other solution adopted with try/catch block around .at use. If no section is found an exception, catched, will be raised. See the documentation if needed to confirm the try/catch block. 
See a8cb305

File: src/libs/antares/study/parameters.cpp (hunk 2 above)
Comment (julza03) [reply src/libs/antares/study/parameters.cpp]: OK. Commits b03bbed and b5def8c

File: src/libs/antares/study/parameters.cpp (hunk 1 above)
Comment (julza03) [reply src/libs/antares/study/parameters.cpp]: Partially corrected here : fa68674. 
I've made the correction only for parameter names (but not for sections). Indeed, section names are already checked in the map and I'm not sure that verifying it before will make any difference in the performances of the app. 
I think that this control is useless, even for parameters, with the new implementation (which is not based on the first letter). You may want to remove this function.

File: src/libs/antares/study/parameters.cpp (hunk 3 above)
Comment (julza03) [reply src/libs/antares/study/parameters.cpp]: Ok. Commits 96a1f00 (fix 0a225bf).

**Comment by julza03** [comment]
Other suggestions on this file parameters.cpp : 
- Before merging, you may want to apply a reformat code - I didn't know if my editor was conform to the rules of Antares so I didn't reformatted the code. 
- #include <stdio.h> is the C syntax. To adopt C++ style, it will be better to write : #include <cstdio> (You may check why cstdio is required, because iostream, already included, may answer to your needs. But CLion warns that these library are unused - you may want to do an analysis). 
- In SGDIntLoadFamily_Optimization, the if concerning "link-type" may be rewritten like this : 
if (key == "link-type")
    {
        CString<64, false> v = value;
        v.trim();
        v.toLower();
        if (value == "ac")
            d.linkType = ltAC;
        else
            d.linkType = ltLocal;
        return true;
    }
- As in the previous version, these comments, inLoadFromINI don't reflect the code : // We convert the key and the value into the lower case format. // Do not forget the variable `key` and `value` are identical to
            // `p->key` and `p->value` except they are already in the lower case format
Only the value is converted in lower case, I think that the key must be converted too because the comparisons in SGDIntLoadFamily are made with text in lowercase. (Maybe p->key is already in lower-case, I don't know.)
-saveToINI is a very long function. I think you may want to split it in several ones, each one in charge of saving one section.

File: src/libs/antares/study/parameters.cpp (hunk 1 above)
Comment (flomnes) [reply src/libs/antares/study/parameters.cpp]: I agree

File: src/libs/antares/study/parameters.cpp [hunk 7]
```diff
@@ -1020,24 +925,57 @@ bool Parameters::loadFromINI(const IniFile& ini, uint version, const StudyLoadOp
...
+                continue;
             // We convert the key and the value into the lower case format
             value = p->value;
             value.toLower();
 
             // Deal with the current property
-            if (not ParametersInternalLoadProperty(*this, p->key, value, p->value, version))
+            // Do not forget the variable `key` and `value` are identical to
+            // `p->key` and `p->value` except they are already in the lower case format
+            if (not (handleAllKeysInSection)(*this, p->key, value, p->value, version))
```
Comment (flomnes) [inline src/libs/antares/study/parameters.cpp]: Style : parenthesis around function name are not needed.

File: src/libs/antares/study/parameters.cpp [hunk 8]
```diff
@@ -1020,24 +925,57 @@ bool Parameters::loadFromINI(const IniFile& ini, uint version, const StudyLoadOp
     reset();
     // A temporary buffer, used for the values in lowercase
     String value;
-
+    String sectionName;
+    typedef bool (*Callback)(Parameters&, const String&, const String&, const String&, uint);
```
Comment (flomnes) [inline src/libs/antares/study/parameters.cpp]: For clarity, I suggest you comment each parameter
Example :
```cpp
typedef bool (*Callback)(Parameters& // [out] Parameter object to load the data into
                         const String& // [in] Key, comes left to the '=' sign in the .ini file
                         const String& // [in] Lowercase value, comes right to the '=' sign in the .ini file
                                       // etc.
                        );

**Review by flomnes — COMMENTED** [review]
Formatting may be incorrect, I'll fix it before a merge.

To be added to Antares-next

File: src/libs/antares/study/parameters.cpp (hunk 8 above)
Comment (julza03) [reply src/libs/antares/study/parameters.cpp]: Ok : 0b051f3

File: src/libs/antares/study/parameters.cpp (hunk 7 above)
Comment (julza03) [reply src/libs/antares/study/parameters.cpp]: Ok : 0b051f3

File: src/libs/antares/study/parameters.cpp [hunk 9]
```diff
@@ -926,7 +926,11 @@ bool Parameters::loadFromINI(const IniFile& ini, uint version, const StudyLoadOp
     // A temporary buffer, used for the values in lowercase
     String value;
     String sectionName;
-    typedef bool (*Callback)(Parameters&, const String&, const String&, const String&, uint);
+    typedef bool (*Callback)(Parameters&, // [out] Parameter object to load the data into
+            const String&, // [in] Key, comes left to the '=' sign in the .ini file
+            const String&, // [in] Lowercase value, comes right to the '=' sign in the .ini file
+            const String&, // [in] Raw value as writtent right to the '=' sign in the .ini file
+            uint); // [in] Version of the simulator (such as 710)
```
Comment (flomnes) [inline src/libs/antares/study/parameters.cpp]: Upon closer examination, the version is read from the **study**, it is not the version of Antares Simulator.

**load.cpp**
```cpp
bool Study::internalLoadHeader(const String& path)
{
    // Header
    buffer.clear() << path << SEP << "study.antares";
    if (not header.loadFromFile(buffer))
    {
        logs.error() << path << ": impossible to open the header file";
        return false;
    }
// ...
}
```
**header.cpp**
```cpp
bool StudyHeader::internalLoadFromINIFile(const IniFile& ini, bool warnings)
{
    version = 0;
    const IniFile::Section* sect = ini.find("antares");
// ...
            // Version
            if (p->key == "version")
            {
                version = p->value.to<uint>(); // Property "version" in section "antares" from file study.antares
                continue;
            }
// ...
}
```

**Review by flomnes — COMMENTED** [review]
One last change: comment at line 933

**Review by guilpier-code — APPROVED** [review]

**Review by flomnes — APPROVED** [review]

**Pull request merged**